src/Workspaces/Core/Portable/Workspace/Solution/SolutionCompilationState.cs

```diff
@@ -710,19 +710,34 @@ public SolutionCompilationState RemoveAnalyzerReference(ProjectId projectId, Ana
 
     /// <inheritdoc cref="SolutionState.WithDocumentText(DocumentId, SourceText, PreservationMode)"/>
     public SolutionCompilationState WithDocumentText(DocumentId documentId, SourceText text, PreservationMode mode)
-        => WithDocumentTexts([(documentId, text, mode)]);
+    {
+        // before allocating an array below, do a fast check if the text has actually changed. this shows up in
+        // allocation traces and is worthwhile to avoid for the common case where we're continually being asked to
+        // update the same doc to the same text (for example, when GetOpenDocumentInCurrentContextWithChanges) is
+        // called.
+        //
+        // The use of GetRequiredState mirrors what happens in WithDocumentTexts
+        var documentState = this.SolutionState
+            .GetRequiredProjectState(documentId.ProjectId).DocumentStates
+            .GetRequiredState(documentId);
+        if (SourceTextIsUnchanged(documentState, text))
+            return this;
+
+        return WithDocumentTexts([(documentId, text, mode)]);
```

Comment:
>WithDocumentTexts [](http://example.com/codeflow?start=15&length=17)

I assume most callers of WithDocumentTexts outside this one actually do have a modification? (Ie, it's not worth lifting the check inside WithDocumentContents outside where the allocations happen in that method)

Reply:
yeah, there's only one caller of this.  so i can life to there.

Reply:
done.